pkg/collector/corechecks/servicediscovery/servicediscovery.go

```diff
@@ -128,7 +153,124 @@ func (c *Check) Configure(senderManager sender.SenderManager, _ uint64, instance
 
 // Run executes the check.
 func (c *Check) Run() error {
-	return c.os.DiscoverServices()
+	start := time.Now()
+	defer func() {
+		diff := time.Since(start).Seconds()
+		metricTimeToScan.Observe(diff)
+	}()
+
+	disc, err := c.os.DiscoverServices()
+	if err != nil {
+		telemetryFromError(err)
+		return err
+	}
+
+	log.Debugf("aliveProcs: %d | ignoreProcs: %d | runningServices: %d | potentials: %d | openPorts: %s",
+		disc.aliveProcsCount,
+		len(disc.ignoreProcs),
+		len(disc.runningServices),
+		len(disc.potentials),
+		disc.openPorts.String(),
+	)
+	metricDiscoveredServices.Set(float64(len(disc.runningServices)))
+
+	runningServicesByName := map[string][]*serviceInfo{}
+	for _, svc := range disc.runningServices {
+		runningServicesByName[svc.meta.Name] = append(runningServicesByName[svc.meta.Name], svc)
+	}
+	for _, svcs := range runningServicesByName {
+		if len(svcs) <= 1 {
+			continue
+		}
+		for _, svc := range svcs {
+			if c.sentRepeatedEventPIDs[svc.process.PID] {
+				continue
+			}
+			err := fmt.Errorf("found repeated service name: %s", svc.meta.Name)
+			telemetryFromError(errWithCode{
+				err:  err,
+				code: errorCodeRepeatedServiceName,
+				svc:  &svc.meta,
+			})
+			// track the PID, so we don't increase this counter in every run of the check.
+			c.sentRepeatedEventPIDs[svc.process.PID] = true
+		}
+	}
+
+	potentialNames := map[string]bool{}
+	for _, p := range disc.potentials {
+		potentialNames[p.meta.Name] = true
+	}
+
+	// group events by name in order to find repeated events for the same service name.
+	eventsByName := make(eventsByNameMap)
```

Comment:
we group by name but that key is not guarantee to be unique. Is the logic correct in this way or should we have a better key?

Reply:
eventsByNameMap has a key of name and a value that's a struct containing slices, each one representing a service state. So if there's a collision, we should see all the processes with the same name in all states.